fdbserver/ClusterController.actor.cpp

```diff
@@ -1783,15 +1785,22 @@ void registerWorker( RegisterWorkerRequest req, ClusterControllerData *self ) {
 		}
 	}
 
-	if ( req.distributorInterf.present() && !self->db.serverInfo->get().distributor.present() ) {
+	if (req.distributorInterf.present() && !self->db.serverInfo->get().distributor.present() &&
+			self->clusterControllerDcId == req.distributorInterf.get().locality.dcId() &&
+			!self->recruitingDistributor) {
 		const DataDistributorInterface& di = req.distributorInterf.get();
 		TraceEvent("ClusterController_RegisterDataDistributor", self->id).detail("DDID", di.id());
 		self->db.setDistributor(di);
 	}
-	if ( req.ratekeeperInterf.present() && !self->db.serverInfo->get().ratekeeper.present() ) {
-		const RatekeeperInterface& rki = req.ratekeeperInterf.get();
-		TraceEvent("ClusterController_RegisterRatekeeper", self->id).detail("RKID", rki.id());
-		self->db.setRatekeeper(rki);
+	if (req.ratekeeperInterf.present() && !self->recruitingRatekeeper) {
```

Comment:
this misses ratekeepers which register while we are recruiting, explicitly pass an ID to ratekeeper so we are able to detect the worker is the one we recruited